### PR TITLE
Add Experimental Preset Support

### PR DESCRIPTION
**Pull request by TheLarkInn**
**What kind of change does this PR introduce?**
Feature (WIP)

**Did you add tests for your changes?**
No this is experimental right now. Kind of WIP to also discuss implementation details. 

**If relevant, link to documentation update:**

**Summary**
The following PR adds a Work in Progress, Proof of Concepts of a highly primitive preset system. 

A webpack preset is detected by searching a users "package.json" devDependencies for all package names starting with `webpack-preset-`. From there we will require those dependencies and they will yield their configuration objects which are `webpack-merge`'d together. 

Whether or not we think #0CJS is a big deal, extending it is what it really means to empathize with a large subset of our users we have yet to tap into. Also, I would like this PR to be a place to help start the design discussion that is testable. 

A preset is just a module that exports a configuration. <==== **likely where we want to design more**

**Does this PR introduce a breaking change?**
Nope.

**Other information**

### CODE REVIEW

**Comment by clarkio** [comment]
Code reviewed on twitch. Good to go 👍  😜

**Comment by ooflorent** [comment]
**tl;dr:** Some features I would expect to be supported by webpack presets:

- Scoped packages
- Resolution order of presets
- Environment specific presets

---

Presets are good for 0CJS but there are a few things that won't work with your proposal.

> A webpack preset is detected by searching a users "package.json" devDependencies

Presets shouldn't be restricted to `devDependencies`. A newcomer may unintentionally install a preset without `--dev` and may have a hard time finding why it is not working as expected.

If the presets are discovered by reading the dependencies, then presets included by another packages won't be found.

>  package names starting with `webpack-preset-`.

We should limit to this because of scoped packages.

Dependencies are usually alphabetically sorted. Which means that _preset A_ will be applied before _preset B_. Maintaining the desired order would be a pretty bad experience for the user since package managers could override/re-sort the dependencies.

Applying all presets found in `package.json` means that it is not possible to skip some depending on your config (`target`, `mode`, …). Babel solved this by having `env` and reading `.babelrc.js` (so you could programmatically define what's exported/used). Webpack already has config files. Using them would solve the issue but it's not longer 0CJS.

---

Regarding the implementation, webpack shouldn't use `webpack-merge` but it's own module/package. Depending on a 3rd party dependency would complicate how developers evolve webpack.

`cwd` is not reliable for discovery. For example, in a mono-repository, a developer could invoke the build using: `packages/some/build.sh`. `cwd` would be the root but the read `package.json` should be `packages/some/package.json`.

File: lib/webpack.js [hunk 1]
```diff
@@ -12,8 +12,10 @@ const WebpackOptionsDefaulter = require("./WebpackOptionsDefaulter");
 const validateSchema = require("./validateSchema");
 const WebpackOptionsValidationError = require("./WebpackOptionsValidationError");
 const webpackOptionsSchema = require("../schemas/WebpackOptions.json");
+const WebpackPresetsApply = require("./WebpackPresetsApply");
 
 const webpack = (options, callback) => {
+	console.log("WERBPERK");
```
[the remark this answers is not in the document]
Comment (TheLarkInn) [reply lib/webpack.js]: Nope this is still proof of concept and is a good reminder that I forget console logs. 😬

**Comment by TheLarkInn** [comment]
Thanks for the comments @ooflorent

So each of these things you've mentioned we're explicitly not included so it could spur this discussion. I will take some time this evening and comment on possible design choices. 

I do however, think that leveraging webpack-merge is a good idea.

**Comment by donaldpipowitch** [comment]
While I like every way to make and share configuration easier I'd like to see a more general approach to composing, updating and simplifying the webpack config. Loading presets on a namespace-based pattern seems to b a little bit limited. Can I for example use different webpack builds/configs in one project with presets (because I only have one package.json and I don't want to merge _every_ preset)? E.g. I use webpack to build my application, my unit tests, my E2E tests and maybe more depending in the project.

So maybe my setup looks kinda like this:

```
src/   # application code
tests/ # unit tests
e2e/   # e2e tests
webpack.src.config.js
webpack.tests.config.js
webpack.e2e.config.js
```
or 
```
src/webpack.config.js
tests/webpack.config.js
e2e/webpack.config.js
```

Webpack everything 💕

**Comment by eliperelman** [comment]
> Presets are good for 0CJS

I agree, and I also think using presets are more important than trying to say the whole thing is zero-config. In Neutrino, we give users two options:

1. True zero-config via CLI

```sh
neutrino start --use @neutrinojs/airbnb @neutrinojs/react
```

2. No-config but "setup", kinda like declaring your dependencies. You don't need to configure them, you just have to tell Neutrino which ones you want to use:

```sh
neutrino start
```

```js
// .neutrinorc.js
module.exports = {
  use: [
    '@neutrinojs/airbnb',
    '@neutrinojs/react'
  ]
};
```

> A newcomer may unintentionally install a preset 

This is also true of transitive dependencies that may affect how the project is built and silently modify configuration. Declaring dependencies explicitly avoids this.

---

> webpack shouldn't use `webpack-merge` but it's own module/package

This is something I've harped on before, but this is the bigger issue of composability of presets.

1. If `preset-b` is loaded after `preset-a` and wants to modify the Babel configuration, how can it deterministically know the correct location of the Babel configuration? This not only includes ordering, but also has to deal with the multiple ways that webpack can define these structures.

2. How can I modify the arguments to a plugin without redeclaring the whole plugin and all its options already set?

These are problems that are mostly solved with a different configuration API. For Neutrino we built an alternative API called [webpack-chain](https://github.com/mozilla-neutrino/webpack-chain). I'd be happy to discuss the ergonomics of the API, and I'd even consider donating it to webpack if it has value.

---

> `cwd` is not reliable for discovery.

Agreed. This need to be an option for the tool. In Neutrino, we allow this with the `root` option in `.neutrinorc.js`, and when using the low-level API:

```js
// .neutrinorc.js
module.exports = {
  options: {
    root: __dirname
  }
};
```

```js
// API usage in .eslintrc.js
module.exports = Neutrino({ root: __dirname })
  .use('.neutrinorc')
  .call('eslintrc');
```

---

> Another pain point could be incompatibilities between presets. Since plugins are not deduped, presets could apply the same plugin more than once or introduce clashes between rules/resolvers. Debugging this could be hard.

This could be solved by an alternative config API. Using named locations would either cause options to merge or overwrite at that location, depending on what action the preset is taking.

---

I know a refer to a lot of prior art by Neutrino and webpack-chain in this space, but it's mostly because we have spent 8 major versions working on solving these exact problems and want to contribute our lessons learned. If that means deprecating Neutrino and incorporating these directly into webpack, I'd be happy to see this accessible to everyone.

**Comment by TheLarkInn** [comment]
webpack-merge has historically done a really great job at composition, I'll likely spec this out again in a github issue based off of feedback and this prototype, etc.

**Comment by michael-ciniawsky** [comment]
Preset Composition is likely to be very complicated and always will be imho. This would require an intermediate step like `neutrino`/`webpack-blocks` and friends are doing atm by defining composable blocks with a 'extra' API. A preset could  also be 'just' a shareable set of  `webpack.config.js`, which by provides 'sane'  defaults for a certain framework etc (e.g `webpack-preset-react`) and additional optional configs which add support for e.g `HMR/SSR` via config flags (webpack CLI)

### `Usage`

```bash
npm i -D webpack-cli webpack webpack-preset-react
```

```
# 1. Loads webpack-preset-react, 
# 2. Installs dependencies, 
# 3. Stores the React webpack Config in the 'webpack CLI Config Cache' (see below 'why')
# 4. Generates a webpack.config.js in './'
webpack init --preset react
```
```
# 1. Loads the React Config and applies SSR support
# 2. Generates (updated) webpack.config.js in './'
webpack preset react --ssr
```
```
# Always uses the latest generated config by default
webpack
```
```
# Set whatever config and set it to 'latest/current'
webpack config react

webpack config --list

# Available Configurations:
# react (latest)
# react-ssr
```

```
# Every preset supports certain handy CLI commands for project specifics
# The CLI should do that from an 'internal' Config format (re) generating the config on changes

webpack entry add './Foo.js'
webpack entry set app --files 'a.js' 'b.js' 'c.js'
webpack entry rm './App.js

webpack output --path  '...'
webpack output --filename '[name].js'

# and so on...
```

I would personally avoid merging presets at all. Their could be a an additional type for `loaders/plugins` config etc, but imho a preset is 

- 1. Install 
- 2. Init 
- 3. Run

with a few Commands from the CLI for setting `entries`, `output`, `context`.
Alternatively  a handwritten `webpack.config.js` or 'blocks' like `plugins.js`, `modules.js` could be  'merged/applied' with a preset when the config is generated by the CLI

**Comment by michael-ciniawsky** [comment]
### `Preset (naive)`

```
|– webpack.config.js
|– package.json
```

**webpack.config.js**
```js
const entry = (cli) => {
  if (cli) {
    // merge, assign, remove...

    return entries
  }

  return defaults
}

export default const config = (cli) => ({
   entry: cli.entry ? entry(cli) : entry()
   ...
})
```

**package.json**
```js
{
   "name": "webpack-preset-*"
   "version": "1.0.0",
   "main": "webpack.config.js",
   "dependencies": { // installed on webpack init --preset
       "css-loader": "1.0.0",
       "file-loader": "1.0.0"
   },
   "devDependencies": {
     "webpack-preset": "1.0.0"  // eventual Preset API
   }
}
```

**Comment by izaakschroeder** [comment]
I've spent some time dealing with managing a set of "shared" webpack configurations used on several projects. The approach I finally settled on is documented here: https://medium.com/making-internets/86c325a6927f. While I like `webpack-merge` I've found it to be too imprecise for a lot of transformations I wanted to do. Function composition has always been simple, and requires no extra API like `webpack-chain` or `webpack-blocks`, though helper functions are always possible. Anyway, just thought I'd throw my 2¢ in. Carry on the good fight 😄

**Comment by michael-ciniawsky** [comment]
> Function composition has always seemed like a simple mechanism for this, and requires no extra API like `webpack-chain` or `webpack-blocks`.

👍 

A preset should help building the initial boilerplate and provide good defaults ('best practices') etc, while a few (likely project specific) settings (e.g `entry`, `output`, `context`) can be set via the CLI. If additional more 'complex' settings (loaders/plugins with options) are needed write them into a separate file e.g `./.webpack/plugins.js` and let the CLI replace/add this 'block'. It's also the question where/when to 'stop' and one should just use it's own config at this point of the journey then imho :).

**Comment by eliperelman** [comment]
> Function composition has always seemed like a simple mechanism for this

Completely agree, and this is mostly the approach that Neutrino takes on top of webpack-chain. Unfortunately the problems we are trying to solve with the webpack config object API are not solved via composition, this mostly just aids in breaking up that config into discreet chunks. There is still non-determinism in dealing with loader order and structure. The reason the solution works well for you is you are still in control of the entire configuration and have knowledge about the structure, so you can modify it easily.

Now imagine that the internals of these presets are managed externally, and you want to add a Babel plugin to babel-loader. You know you can add your own function to the pipeline and modify the configuration, but in this object, how do you know where to inject your plugin at? What if you are using a preset that comes with babel-minify-webpack-plugin, and you want to shut off mangling, but want all the other options to stay the same? Should you be required to re-configure that plugin yourself? You can't do that right now with webpack's object config syntax.

> while a few (likely project specific) settings (e.g `entry`, `output`, `context`)

In practice I have found that some of these options can be quite intertwined. Multi-page apps, hot module replacement, dev server; some of these values rely on each other. If you decide that you want to change these values, then precedence has to be taken into account, as these kinds of options need to be set first in the pipeline, while just adding them to the end of the pipeline would not have the desired effect.

> It's also the question where/when to 'stop' and one should just use it's own config at this point of the journey then imho

I'd want to say that even with significant overrides or additions to a *fully-featured* preset, it's still preferable to having to start over. Maybe your preset handles a lot, and you want to change the Babel config, add postcss, change up some linting rules, but don't want to lose the style and extraction bits.

**Comment by michael-ciniawsky** [comment]
> Unfortunately the problems we are trying to solve with the webpack config object API are not solved via composition, this mostly just aids in breaking up that config into discreet chunks.

True, the 'how' for the discrete chunks is the difficult part. I consciously left that out for initial discussion in my comments

> Now imagine that the internals of these presets are managed externally, and you want to add a Babel plugin to babel-loader.

`.babelrc`, `postcssrc.js`, `posthtmlrc.js` etc. ? And this also tackles the difficult parts already (Rules, Loaders with Options, Plugins with Options etc) and if one wants to customize all that why a preset in the first place and not just a base config (`webpack init`) (maybe include links to the some initial docs there) and presets are not for you ?

```
|– .webpack // webpack CLI (don't touch this folder, .webpack for e.g :))
|   |– webpack.config.js // this will run
|– src
|
|– webpack.config.js* // my custom stuff (optional)
|– package.json
```

**\*webpack.config.js**
```js
export const extend = {
   rules: [
     // my custom rules...
   ],
   plugins: [
     // my custom plugins...
   ]
}
```

Yes the strategy (API) how to apply/merge/override custom rules, plugins etc. still needs to be found here and the `custom.config.js` should not be named `webpack.config.js` nor be a plain webpack config at best :). Here something like `webpack-chain` etc might be come in handy, something the CLI can 'understand' on how to proceed :) `rules.prepend(...rule).override(...rule).append(..rule)`

> In practice I have found that some of these options can be quite intertwined. Multi-page apps, hot module replacement, dev server; some of these values rely on each other. 

A Preset API or the CLI would need to apply/update e.g `context` to all properties where it is relevant for and which are present in the current config. e.g `const CONTEXT = cli.context || default' 

> I'd want to say that even with significant overrides or additions to a fully-featured preset, it's still preferable to having to start over.

I'm not saying one needs to start completely over, but if your config has diverged from a preset to a certain degree over time, when is it better to eventually 'just' remove the preset dependency and don't expect webpack (CLI) to update/manage your current config within the 'preset boundary' like an eventual update to the preset anymore. (You graduated from that preset and handle stuff on your own at this point)

**Comment by izaakschroeder** [comment]
> Now imagine that the internals of these presets are managed externally, and you want to add a Babel plugin to babel-loader. You know you can add your own function to the pipeline and modify the configuration, but in this object, how do you know where to inject your plugin at? What if you are using a preset that comes with babel-minify-webpack-plugin, and you want to shut off mangling, but want all the other options to stay the same? Should you be required to re-configure that plugin yourself? You can't do that right now with webpack's object config syntax.

I've done something kind of similar before. I've needed to swap the `target` field of `@babel/preset-env` depending on if it's a `node` or `web` build. But you're right in the sense there's some degree of opaqueness to all this, since modifying the babel pipeline is mostly handled by `babel`. But if you use functions, this means you have access to babel's underlying machinery for loading presets and you can walk the plugin tree just as you please.

```js
import {loader} from 'webpack-partial';
import {OptionManager} from '@babe/core';

const babel = manager.init({
  babelrc: true,
  filename: 'somefile.js',
});

const existing = JSON.parse(fs.readFileSync('.babelrc'));

const tweakBabelLoader = loader.update(
  ({loader}) => /babel-loader/.test(loader),
  (loader) => {
    return {
      ..loader,
      options: {
        plugins: babel.plugins.map((plugin) => { 
          /* mangle plugin here as desired */ 
        }),
       }
    }
);

export default tweakBabelLoader;
```

File: lib/WebpackPresetsApply.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+const path = require("path");
+const PresetsApply = require("./PresetsApply");
+
+const pkgPath = path.join(process.cwd(), "package.json");
+const pkg = require(pkgPath);
+const presets = Object.keys(pkg.devDependencies)
+	.filter(pkgName => pkgName.startsWith("webpack-preset-"))
+	.map(pkgName => path.resolve(process.cwd(), "node_modules", pkgName))
```
Comment (opl-) [inline lib/WebpackPresetsApply.js]: This approach will not work when the presets are hoisted up somewhere higher in the directory structure, are installed globally or when Yarn PnP is used. Maybe enhanced-resolve could be used instead?

**Review by invisiblepancake — APPROVED** [review]

**Comment by alexander-akait** [comment]
Close in favor of https://github.com/webpack/webpack/pull/17171
